vendor/github.com/AudriusButkevicius/pfilter/conn.go

```diff
@@ -0,0 +1,93 @@
+package pfilter
+
+import (
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+type FilteredConn struct {
+	source   *PacketFilter
+	priority int
+
+	recvBuffer chan packet
+
+	filter Filter
+
+	deadline atomic.Value
+	closed   chan struct{}
+}
+
+// LocalAddr returns the local address
+func (r *FilteredConn) LocalAddr() net.Addr {
+	return r.source.LocalAddr()
+}
+
+// SetReadDeadline sets a read deadline
+func (r *FilteredConn) SetReadDeadline(t time.Time) error {
+	r.deadline.Store(t)
+	return nil
+}
+
+// SetWriteDeadline sets a write deadline
+func (r *FilteredConn) SetWriteDeadline(t time.Time) error {
+	return r.source.SetWriteDeadline(t)
+}
+
+// SetDeadline sets a read and a write deadline
+func (r *FilteredConn) SetDeadline(t time.Time) error {
+	r.SetReadDeadline(t)
+	return r.SetWriteDeadline(t)
+}
+
+// WriteTo writes bytes to the given address
+func (r *FilteredConn) WriteTo(b []byte, addr net.Addr) (n int, err error) {
+	select {
+	case <-r.closed:
+		return 0, errClosed
+	default:
+	}
+
+	if r.filter != nil {
+		r.filter.Outgoing(b, addr)
+	}
+	return r.source.WriteTo(b, addr)
+}
+
+// ReadFrom reads from the filtered connection
+func (r *FilteredConn) ReadFrom(b []byte) (n int, addr net.Addr, err error) {
+	select {
+	case <-r.closed:
+		return 0, nil, errClosed
+	default:
+	}
+
+	var timeout <-chan time.Time
+
+	if deadline, ok := r.deadline.Load().(time.Time); ok && !deadline.IsZero() {
+		timeout = time.After(deadline.Sub(time.Now()))
```

Comment:
Will most of our calls have a read deadline or no? If no, this if fine. If yes, we should probably not spawn a goroutine for every read.